internal/cli/login/k8s.go

```diff
@@ -0,0 +1,133 @@
+	// Corresponds to the header expected by basic-auth challenging authenticators
+	// copied from github.com/openshift/library-go/pkg/oauth/tokenrequest/request_token.go
+	csrfTokenHeader = "X-CSRF-Token" //nolint:gosec
+)
+
+type K8sOauth struct {
+	ClientId           string
+	CAFile             string
+	InsecureSkipVerify bool
+	ConfigUrl          string
+}
+
+func NewK8sOAuth2Config(caFile, clientId, authUrl string, insecure bool) K8sOauth {
+	return K8sOauth{
+		CAFile:             caFile,
+		InsecureSkipVerify: insecure,
+		ClientId:           clientId,
+		ConfigUrl:          fmt.Sprintf("%s/.well-known/oauth-authorization-server", authUrl),
+	}
+}
+
+func (k K8sOauth) getOAuth2Client(callback string) (*osincli.Client, error) {
+	oauthServerResponse, err := getOAuth2Config(k.ConfigUrl, k.CAFile, k.InsecureSkipVerify)
+	if err != nil {
+		return nil, err
+	}
+
+	redirectUrl := callback
+	if redirectUrl == "" {
+		redirectUrl = oauthServerResponse.TokenEndpoint + "/implicit"
+	}
+
+	config := &osincli.ClientConfig{
+		ClientId:           k.ClientId,
+		AuthorizeUrl:       oauthServerResponse.AuthEndpoint,
+		TokenUrl:           oauthServerResponse.TokenEndpoint,
+		ErrorsInStatusCode: true,
+		RedirectUrl:        redirectUrl,
+	}
+
+	client, err := osincli.NewClient(config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create oauth2 client: %w", err)
+	}
+
+	tlsConfig, err := getAuthClientTlsConfig(k.CAFile, k.InsecureSkipVerify)
+	if err != nil {
+		return nil, err
+	}
+	client.Transport = &http.Transport{
+		TLSClientConfig: tlsConfig,
+	}
+
+	return client, nil
+}
+
+func (k K8sOauth) authHeadless(username, password string) (AuthInfo, error) {
+	ret := AuthInfo{}
+	client, err := k.getOAuth2Client("")
+	if err != nil {
+		return ret, err
+	}
+	authorizeRequest := client.NewAuthorizeRequest(osincli.CODE)
+	requestURL := authorizeRequest.GetAuthorizeUrl().String()
+	req, err := http.NewRequest(http.MethodGet, requestURL, nil)
+	if err != nil {
+		return ret, err
+	}
+	req.Header.Set(csrfTokenHeader, "1")
+	req.Header.Set("Authorization", "Basic "+base64.StdEncoding.EncodeToString([]byte(username+":"+password)))
+	resp, err := client.Transport.RoundTrip(req)
+
+	if err != nil {
+		return ret, err
+	}
+
+	if resp.StatusCode == http.StatusFound {
+		redirectURL := resp.Header.Get("Location")
+
+		req, err := http.NewRequest(http.MethodGet, redirectURL, nil)
+		if err != nil {
+			return ret, err
+		}
+
+		return getOAuth2AccessToken(client, authorizeRequest, req)
+	}
+	return ret, fmt.Errorf("unexpected http code: %v", resp.StatusCode)
+}
+
+func (k K8sOauth) Auth(web bool, username, password string) (AuthInfo, error) {
+	if web {
+		return oauth2AuthCodeFlow(k.getOAuth2Client)
+	}
+	return k.authHeadless(username, password)
+}
+
+func (k K8sOauth) Renew(refreshToken string) (AuthInfo, error) {
+	return oauth2RefreshTokenFlow(refreshToken, k.getOAuth2Client)
+}
+
+func (k K8sOauth) Validate(args ValidateArgs) error {
```

Comment:
Is this only called from CLI?  Are these params also relevant for device simulator?

Reply:
Yes, only CLI. The devicesimulator uses previously created config.yaml.